cpu/nrf52/periph/adc.c

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright (C) 2017 HAW Hamburg
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     cpu_nrf52
+ * @{
+ *
+ * @file
+ * @brief       Low-level ADC driver implementation
+ *
+ * @author      Dimitri Nahm <dimitri.nahm@haw-hamburg.de>
+ *
+ * @}
+ */
+
+#include "cpu.h"
+#include "mutex.h"
+#include "periph/adc.h"
+#include "periph_conf.h"
+
+#ifdef ADC_CONFIG
+/**
+ * @brief   Load the ADC configuration
+ */
+static const uint8_t adc_config[] = ADC_CONFIG;
+
+/**
+ * @brief   Lock to prevent concurrency issues when used from different threads
+ */
+static mutex_t lock;
+
+static inline void prep(void)
+{
+    mutex_lock(&lock);
+    NRF_SAADC->ENABLE = 1;
+}
+
+static inline void done(void)
+{
+    NRF_SAADC->ENABLE = 0;
+    mutex_unlock(&lock);
+}
+
+int adc_init(adc_t line)
+{
+    if (line >= ADC_NUMOF) {
+        return -1;
+    }
+    return 0;
+}
+
+int adc_sample(adc_t line, adc_res_t res)
+{
+    int val;
+
+    /* check if resolution is valid */
+    if (res > 3) {
+        return -1;
+    }
+
+    /* prepare device */
+    prep();
+
+    /* set line, resolution and use VDD (+5V) as reference */
+    NRF_SAADC->CH[0].PSELP = adc_config[line] + 1;
+    NRF_SAADC->CH[0].CONFIG = (SAADC_CH_CONFIG_GAIN_Gain1_4 << SAADC_CH_CONFIG_RESN_Pos) |
+                              (SAADC_CH_CONFIG_REFSEL_VDD1_4 << SAADC_CH_CONFIG_REFSEL_Pos);
+    NRF_SAADC->RESOLUTION = res - 1;
```

Comment:
this could lead to a negative res value (in case `res := 0`)? Why not override `adc_res_t` for the CPU, so that it takes exactly the values allowed/needed?

Reply:
@haukepetersen Yes, of course.

Reply:
Also here...